### ISSUE TITLE
Installing fasttext with skift doesn't work

### ISSUE THREAD

**Issue by polm** [shaypal5/skift#1]
Tried running this from the README:

    pip install skift[fasttext] --process-dependency-links

Got this error:

    Collecting fasttext==0.1.0+git.3b5fd29; extra == "fasttext" (from skift[fasttext])
      Could not find a version that satisfies the requirement fasttext==0.1.0+git.3b5fd29; extra == "fasttext" (from skift[fasttext]) (from versions: 0.2.0, 0.2.1, 0.3.0, 0.3.1, 0.4.0, 0.5.0, 0.5.1, 0.5.12, 0.5.13, 0.5.14, 0.5.15, 0.5.16, 0.5.17, 0.5.18, 0.5.19, 0.6.0, 0.6.1, 0.6.2, 0.6.4, 0.7.0, 0.7.1, 0.7.2, 0.7.3, 0.7.4, 0.7.5, 0.7.6, 0.8.0, 0.8.1, 0.8.2, 0.8.3)
     No matching distribution found for fasttext==0.1.0+git.3b5fd29; extra == "fasttext" (from skift[fasttext])

Tried with Python 3.6.4 in and out of a virtualenv. Seems skift expects to find a version of fasttext that's not available in pypi?

**Comment by shaypal5** [comment]
That's true. The script specifies [a specific commit](https://github.com/facebookresearch/fastText/commit/3b5fd293597de550a131d81436b9755902f39bb2) of [the official `fasttext` repository](https://github.com/facebookresearch/fastText) as the target for the `fasttext` dependency.

That dependency is detailed using a dependency link, hence the obligatory use of the `--process-dependency-links` flag.

I'm not sure what might be the problem, as this installation method is tested continuously through travis. You can see the tests (for both Python 3.5 and Python 3.6) for the specific version of the code bundled as `skift v0.0.6` here:
https://travis-ci.org/shaypal5/skift/jobs/340472996

Can you check your pip version using the `pip --version` command and report it here?

Basically, these are the relevant `pip` prints:
```Collecting fasttext==0.1.0+git.3b5fd29 (from skift==0.0.6)
Cloning https://github.com/facebookresearch/fastText.git (to 3b5fd293597de550a131d81436b9755902f39bb2) to /tmp/pip-build-ql5i7fzc/fasttext
Could not find a tag or branch '3b5fd293597de550a131d81436b9755902f39bb2', assuming commit.
Requested fasttext==0.1.0+git.3b5fd29 from git+https://github.com/facebookresearch/fastText.git@3b5fd293597de550a131d81436b9755902f39bb2#egg=fasttext-0.1.0+git.3b5fd29 (from skift==0.0.6), but installing version None
DEPRECATION: Dependency Links processing has been deprecated and will be removed in a future release.```

**Comment by shaypal5** [comment]
I'm also wondering why your pip isn't writing the version of `skift` you are trying to install. As you can see in my quote above, it should write `(from skift==0.0.6)` after each dependency it processes.

That's true for every package installation that I encountered, generally. At least with `pip 9.0.1`.

**Comment by polm** [comment]
`pip --version` is 9.0.1; I am on Arch Linux and using the current package (9.0.1-3). The output didn't mention skift because I clipped it to just the error part, here's the whole thing copy/pasted from an install in a fresh virtualenv (though the system download cache is being reused).

    > pip install skift[fasttext] --process-dependency-links
    Collecting skift[fasttext]
      Using cached skift-0.0.6-py2.py3-none-any.whl
    Collecting scipy (from skift[fasttext])
      Using cached scipy-1.0.0-cp36-cp36m-manylinux1_x86_64.whl
    Collecting numpy (from skift[fasttext])
      Using cached numpy-1.14.0-cp36-cp36m-manylinux1_x86_64.whl
    Collecting scikit-learn (from skift[fasttext])
      Using cached scikit_learn-0.19.1-cp36-cp36m-manylinux1_x86_64.whl
    Collecting fasttext==0.1.0+git.3b5fd29; extra == "fasttext" (from skift[fasttext])
      Could not find a version that satisfies the requirement fasttext==0.1.0+git.3b5fd29; extra == "fasttext" (from skift[fasttext]) (from versions: 0.2.0, 0.2.1, 0.3.0, 0.3.1, 0.4.0, 0.5.0, 0.5.1, 0.5.12, 0.5.13, 0.5.14, 0.5.15, 0.5.16, 0.5.17, 0.5.18, 0.5.19, 0.6.0, 0.6.1, 0.6.2, 0.6.4, 0.7.0, 0.7.1, 0.7.2, 0.7.3, 0.7.4, 0.7.5, 0.7.6, 0.8.0, 0.8.1, 0.8.2, 0.8.3)
    No matching distribution found for fasttext==0.1.0+git.3b5fd29; extra == "fasttext" (from skift[fasttext])

Note that I can get skift working without issue by checking out fastText and installing like `pip install /path/to/fasttext`, it's just this pip install of fasttext that doesn't work.

**Comment by shaypal5** [comment]
Ok. So the fact that the command doesn't output the following:
`DEPRECATION: Dependency Links processing has been deprecated and will be removed in a future release.`
Tells us something. It's not even trying to process dependency links.

I'm afraid it might be the case that the square brackets, which are  a bash operator at least in some contexts, might be swallowing the following flag.

Can you please try running
`pip install "skift[fasttext]" --process-dependency-links`
And tell me if it works?

Cheers,
Shay

**Comment by polm** [comment]
I had the thought the brackets might be a problem too, and tried quoting it. Nothing changed though.

I also tried putting the `--process-dependency-links` before `skift[fasttext]`, but again nothing changed.

Just in case I also tried using the `--no-cache-dir` option, but that also had no effect.

**Comment by shaypal5** [comment]
Hmmm. Maybe you can somehow set to never process dependency links in you `~/.pip/pip.conf` file? I'm just guessing wildly here.

**Comment by polm** [comment]
I don't have a `pip.conf` file, not even in `/etc` :/ Sorry I have no idea what's going on...

Do you know another package that uses `--process-dependency-links`? If so I can see if it's any different; if it's not I can ask about the general problem on StackOverflow or something.

**Comment by shaypal5** [comment]
OK. After some local testing I've arrived to the conclusion that dependency links are not a reliable mechanism in `pip` (especially since they been taken out of the tool and then returned as a deprecated feature).

As such, I'm letting go of the idea of providing a way to install `fasttext` as a dependency of `skift`. Instead, I'm going to always be testing `skift` against a specific commit of the official `fasttext` repository - moving to a newer commit periodically - and write on how to easily install that version of `fasttext` in my README file.

Please check out the new README and tell me what you think. Specifically, I'd also love you to test this method yourself, if you can spare a minute (although it is now the way I install `fasttext` in my travis build).

I'm waiting on your input to close this issue.

Cheers,
Shay

**Comment by polm** [comment]
Thanks, the new instructions in the README worked just fine! Thanks for being so thorough about it.

On a happier note, it looks like the maintainer of `fasttext` on pypi may have changed based on [this commit](https://github.com/pyk/fastText.py/commit/ed4807359d31296f444f895d26200140b8ce10b4#diff-04c6e90faac2675aa89e2176d2eec7d8), so maybe there's hope for having a usable version on pypi soon too.

**Comment by shaypal5** [comment]
Great! I'm closing this issue for now, and I promise to move `skift` to build on top of the PyPI version of `fasttext` as soon as the official version is on it (I agree that it looks like it might already be the case, but I'll wait for an official answer on the issue you opened).   ( :

Feel free to open an issue here the minute I can make this transition.
